### PR TITLE
virt-launcher pod CPU should be set according to the value calculated…

### LINKED ISSUE #7334: cpuAllocationRatio does not take effect
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
> /kind bug
I configured the value of cpuAllocationRatio in kubevirt CR, but it does not take effect.
```yaml
kind: KubeVirt
metadata:
    name: kubevirt-kubevirt-hyperconverged
    namespace: kubevirt
spec:
    certificateRotateStrategy:
      selfSigned:
        ca:
          duration: 87600h0m0s
          renewBefore: 720h0m0s
        server:
          duration: 87600h0m0s
          renewBefore: 720h0m0s
    configuration:
      developerConfiguration:
        cpuAllocationRatio: 4
        diskVerification:
          memoryLimit: 2G
        featureGates:
        - DataVolumes
        - SRIOV
        - CPUManager
        - CPUNodeDiscovery
        - Snapshot
        - HotplugVolumes
        - ExpandDisks
        - GPU
        - HostDevices
        - DownwardMetrics
        - NUMA
        - WithHostModelCPU
        - HypervStrictCheck
        - SRIOVLiveMigration
      migrations:
        completionTimeoutPerGiB: 800
        parallelMigrationsPerCluster: 5
        parallelOutboundMigrationsPerNode: 2
        progressTimeout: 150
      network:
        defaultNetworkInterface: masquerade
      obsoleteCPUModels:
        "486": true
        Conroe: true
        athlon: true
        core2duo: true
        coreduo: true
        kvm32: true
        kvm64: true
        n270: true
        pentium: true
        pentium2: true
        pentium3: true
        pentiumpro: true
        phenom: true
        qemu32: true
        qemu64: true
      selinuxLauncherType: virt_launcher.process
      smbios:
        family: KubeVirt
        manufacturer: KubeVirt
        product: None
    customizeComponents: {}
    infra:
      replicas: 1
    uninstallStrategy: BlockUninstallIfWorkloadsExist
    workloadUpdateStrategy:
      batchEvictionInterval: 1m0s
      batchEvictionSize: 10
      workloadUpdateMethods:
      - LiveMigrate
```
Then I create the virtual machine, as defined in YAML：
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
metadata:
  name: test-vm
  namespace: test-ns
spec:
  template:
    spec:
      domain:
        cpu:
          cores: 12
        resources:
          limits:
            memory: 2Gi
          requests:
            memory: 2Gi
```
And then , the configuration you see in vmi looks like this:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
metadata:
  name: test-vm
  namespace: test-ns
spec:
  domain:
    cpu:
      cores: 12
      sockets: 1
      threads: 1
      model: host-model
    resources:
      limits:
        memory: 2Gi
      requests:
        memory: 2Gi
```
However, the resource.cpu value of the virt-launcher pod is always 1000m.
```yaml
apiVersion: v1
kind: Pod
metadata:
  name: virt-launcher-test-vm-cnzxg
  namespace: test-ns
spec:
  containers:
  - name: compute
    resources:
      limits:
        cpu: "1"
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
        memory: 2222Mi
      requests:
        cpu: "1"
        devices.kubevirt.io/kvm: "1"
        devices.kubevirt.io/tun: "1"
        devices.kubevirt.io/vhost-net: "1"
        ephemeral-storage: 50M
        memory: 2222Mi
```
But I think, according to Kubevirt's code logic, pod cpu request value should be 12/4=3?
Here is code https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-controller/services/template.go#L883

**What happened**:

cpuAllocationRatio does not take effect

**What you expected to happen**:

cpuAllocationRatio take effect

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`): v0.48.1
- Kubernetes version (use `kubectl version`):  v1.22.3
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): CentOS Linux 7
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by zhhray**
… by the cpuAllocationRatio if it was set greater than 1

**What this PR does / why we need it**:

virt-launcher pod CPU should be set according to the value calculated by the cpuAllocationRatio if it was set greater than 1

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #7334

**Special notes for your reviewer**:

**Release note**:

```release-note

```

### CODE REVIEW

**Comment by zhhray** [comment]
VMI resources may be set by the default value of limitrange. So copy vmi resources to virt-launcher pod firstly.
Then calculate the pod resource.cpu and limit cpu, if the `cpuAllocationRatio` is configured.

File: pkg/virt-controller/services/template.go [hunk 1]
```diff
@@ -878,32 +878,34 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
...
-		if vcpus != 0 && cpuAllocationRatio > 0 {
+		if vcpus != 0 && cpuAllocationRatio > 1 {
 			val := float64(vcpus) / float64(cpuAllocationRatio)
 			vcpusStr := fmt.Sprintf("%g", val)
 			if val < 1 {
 				val *= 1000
 				vcpusStr = fmt.Sprintf("%gm", val)
 			}
 			resources.Requests[k8sv1.ResourceCPU] = resource.MustParse(vcpusStr)
+			resources.Limits[k8sv1.ResourceCPU] = *resource.NewQuantity(vcpus, resource.BinarySI)
```
Comment (xpivarc) [inline pkg/virt-controller/services/template.go:906]: This is interesting. Would you be able to extract this into a separate commit and provide a reason?

File: pkg/virt-controller/services/template.go [hunk 2]
```diff
@@ -878,32 +878,34 @@ func (t *templateService) renderLaunchManifest(vmi *v1.VirtualMachineInstance, i
 	resources.Requests = make(k8sv1.ResourceList)
 	resources.Limits = make(k8sv1.ResourceList)
 
+	// Copy vmi resources requests to a container
```
Comment (xpivarc) [inline pkg/virt-controller/services/template.go:881]: I don't think this is how the `CPUAllocationRatio` is supposed to work. 

I think there are 2 cases:
1. User set a cpu topology. Let's say 1 core.  The expected behavior, in this case, would be to apply the `CPUAllocationRatio`. The resulting VM would not be guaranteed to get the whole cpu time.

2. User wants VM with guaranteed cpu time and therefore sets the `cpu request`. In this case, I don't expect that `CPUAllocationRatio` gets applied.

The problem we need to fix here is to properly distinguish when we are dealing with the 2nd case. User vs limitRange

@vladikr Would you agree?

File: pkg/virt-config/virt-config.go [hunk 3]
```diff
@@ -63,7 +63,7 @@ const (
 	DefaultARCHOVMFPath                             = "/usr/share/OVMF"
 	DefaultAARCH64OVMFPath                          = "/usr/share/AAVMF"
 	DefaultMemBalloonStatsPeriod             uint32 = 10
-	DefaultCPUAllocationRatio                       = 10
+	DefaultCPUAllocationRatio                       = 1
```
[the remark this answers is not in the document]
Comment (zhhray) [reply pkg/virt-config/virt-config.go:66]: In our understanding, if a user has configured the `cpuAllocationRatio`, he must want to enable CPU overcommit for the virtual machine.
So whether set `vmi.spec.domain.cpu` or `vmi.spec.resource.request/limit.cpu`, as long as the configuration `cpuAllocationRatio` in kubevirt cr, that must want the CPU to overcommit. If not, remove the `cpuAllocationRatio` configuration from the Kubevirt CR.
According to the above conditions, then DefaultCPUAllocationRatio == 1, when calculating the CPU of virt-launcher pod, as long as more than 1, some one must have set the `cpuAllocationRatio`.

File: pkg/virt-controller/services/template.go (hunk 2 above)
Comment (zhhray) [reply pkg/virt-controller/services/template.go:881]: Ok, I understand the scenario you're describing.
First consider the starting point I mentioned above. If the `CPUAllocationRatio` is not designed to be what I said.
Please kindly ask whether it(distinguish User vs limitRange) can be implemented in the new version as soon as possible?

File: pkg/virt-config/virt-config.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (vladikr) [reply pkg/virt-config/virt-config.go:66]: Well, we should always overcommit CPU by default to increase the density of pods/VMIs. This is also the default behavior in Kubernetes in general. Each pod by default gets 1/10 of a cputime of a core. If one full core is considered as 1000 mil, the default per pod is 100mil. 
KubeVirt's DefaultCPUAllocationRatio follows k8s and allocated 1/10 of a core for a given pod.

On the other hand, KubeVirt must scale this CPU request with the number of vCPUs the VMI guest requires. 
Without scaling, as the number of vCPUs increases, each will get a lower CPU time when competing with other processes on the node or even with VMIs with a lower amount of vCPUs.
The `cpuAllocationRatio` merely normalizes the requested compute container CPUs based on the total number of vCPUs * cpuAllocationRatio

For example, using the default cpuAllocationRatio and 6 vCPUs, the VMI will look like 

```
spec:
  domain:
    cpu:
      sockets: 1
      cores: 6
      threads: 1
```
The vmipod will request 600mil of cpu time.
```
    resources:
      requests:
        cpu: 600m
```

and 

```
spec:
  domain:
    cpu:
      sockets: 1
      cores: 1
      threads: 1
```
The vmipod will get the default 100mil of cpu time.
```
    resources:
      requests:
        cpu: 100m
```

VMI users can also request guaranteed CPUs and allocate the exact amount of vCPU from the host.
In that case, there will be no overcommit. The vmi pod will be assigned a Guaranteed QoS and will get exclusive CPUs.
This can simply be achieved by requesting a dedicatedCpuPlacement. [1]

```
spec:
  domain:
    cpu:
      sockets: 2
      cores: 1
      threads: 1
      dedicatedCpuPlacement: true
```

If you are simply looking to change the `cpuAllocationRatio`, you can do so by editing the kubevirt CR instead of changing the default.


[1] https://kubevirt.io/user-guide/virtual_machines/dedicated_cpu_resources/#dedicated-cpu-resources

File: pkg/virt-controller/services/template.go (hunk 1 above)
Comment (muff1nman) [reply pkg/virt-controller/services/template.go:906]: Setting a limit, as long as the limit is greater than the request, retains the Burstable QoS, right? https://kubernetes.io/docs/tasks/configure-pod-container/quality-service-pod/#create-a-pod-that-gets-assigned-a-qos-class-of-guaranteed
